lib/rpmsg/rpmsg_virtio.c

```diff
@@ -315,35 +330,52 @@ static int _rpmsg_virtio_get_buffer_size(struct rpmsg_virtio_device *rvdev)
 
 static void rpmsg_virtio_hold_rx_buffer(struct rpmsg_device *rdev, void *rxbuf)
 {
-	struct rpmsg_hdr *rp_hdr;
+	metal_mutex_acquire(&rdev->lock);
+	RPMSG_BUF_HELD_INC(RPMSG_LOCATE_HDR(rxbuf));
+	metal_mutex_release(&rdev->lock);
+}
 
-	(void)rdev;
+static bool rpmsg_virtio_release_rx_buffer_nolock(struct rpmsg_device *rdev,
+						  struct rpmsg_hdr *rp_hdr)
+{
+	struct rpmsg_virtio_device *rvdev;
+	uint16_t idx;
+	uint32_t len;
+
+	rvdev = metal_container_of(rdev, struct rpmsg_virtio_device, rdev);
 
-	rp_hdr = RPMSG_LOCATE_HDR(rxbuf);
+	/* Check the held counter first */
+	if (RPMSG_BUF_HELD_COUNTER(rp_hdr) <= 0)
+		return false;
 
-	/* Set held status to keep buffer */
-	rp_hdr->reserved |= RPMSG_BUF_HELD;
+	/* Decrease the held counter */
+	RPMSG_BUF_HELD_DEC(rp_hdr);
+
+	/* Check whether to release the Rx buffer */
+	if (RPMSG_BUF_HELD_COUNTER(rp_hdr) > 0)
+		return false;
+
+	/* The reserved field contains buffer index */
+	idx = (uint16_t)(rp_hdr->reserved & ~RPMSG_BUF_HELD_MASK);
+	/* Return buffer on virtqueue. */
+	len = virtqueue_get_buffer_length(rvdev->rvq, idx);
+	rpmsg_virtio_return_buffer(rvdev, rp_hdr, len, idx);
+
+	return true;
```

Comment:
regarding the use of this function, I wonder if this should not been split for readability and the buffer count balancing

```c
static bool rpmsg_virtio_buf_held_dec_test(   struct rpmsg_hdr *rp_hdr)
	/* Check the held counter first */
	if (RPMSG_BUF_HELD_COUNTER(rp_hdr) <= 0)
		return false;

	/* Decrease the held counter */
	RPMSG_BUF_HELD_DEC(rp_hdr);

	/* Check whether to release the Rx buffer */
	if (RPMSG_BUF_HELD_COUNTER(rp_hdr) > 0)
		return false;
}

void rpmsg_virtio_release_rx_buffer_nolock(struct rpmsg_virtio_device *rvdev,   struct rpmsg_hdr *rp_hdr)
{
	uint16_t idx;
	uint32_t len;

	/* The reserved field contains buffer index */
	idx = (uint16_t)(rp_hdr->reserved & ~RPMSG_BUF_HELD_MASK);
	/* Return buffer on virtqueue. */
	len = virtqueue_get_buffer_length(rvdev->rvq, idx);
	rpmsg_virtio_return_buffer(rvdev, rp_hdr, len, idx);
}
```

Reply:
Done